File: Problems/No.234_Palindrome_Linked_List/palindrome_linked_list.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def isPalindrome(self, head):
        """
        :type head: ListNode
        :rtype: bool
        """

        # parameters
        front, back, reverse_back = head, head, None

        # start computing (use Floyd's algorithm)

        # find the midium value
        while (back and back.next):
            front = front.next
            back = back.next.next

        # reversal step
        reverse_back = front.next
        front.next = None
        while reverse_back:
            back = reverse_back.next
            reverse_back.next = front
            front = reverse_back
            reverse_back = back

        # comparison step
        back = front
        front = head
        while back:
            if (front.val == back.val):
                front = front.next
                back = back.next
            else:
                return False

        return True

File: Problems/No.234_Palindrome_Linked_List/palindrome_linked_list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    bool isPalindrome(ListNode* head) {
        
        // parameters
        ListNode *front = head, *back = head;
        ListNode *reverse_back;

        // start computing (use Floyd's algorithm)
        
        // find the midium value
        while (back && back->next) {

            front = front->next;
            back = back->next->next;

        }

        // reversal step
        reverse_back = front->next;
        front->next = NULL;
        while (reverse_back) {

            back = reverse_back->next;
            reverse_back->next = front;
            front = reverse_back;
            reverse_back = back;

        }

        // comparison step
        back = front;
        front = head;
        while (back) {

            if (front->val == back->val) {

                front = front->next;
                back = back->next;

            }
            else return false;

        }

        return true;

    }
};
